workspace/src/path_planning/path_planning/path_planning.py

```diff
@@ -207,25 +225,90 @@ def plan_path(self):
+        theta_current = self.state.pose.orientation.z-0.24845347641462115
+        while theta_current<-np.pi:
+            theta_current = theta_current+2*np.pi
+        while theta_current>np.pi:
+            theta_current = theta_current - 2*np.pi
+
+        v_current = np.sqrt(self.state.twist.linear.x**2+self.state.twist.linear.y**2)
+
+        #self.get_logger().info("SPEED: "+str(v_current))
+
+        dist = np.zeros((1,len(self.ref_traj[:,1])))
+        for i in range(len(self.ref_traj[:,1])):
+            dist[0][i] = dist[0][i] = (x_current+np.cos(theta_current)*self.lookahead-self.ref_traj[i][0])**2+(y_current+np.sin(theta_current)*self.lookahead-self.ref_traj[i][1])**2
+        index = dist.argmin()
+
+        ref_state_current = list(self.ref_traj[index,:])
+        err_theta = 0
+        ref = ref_state_current[2]
+        act = theta_current
+
+        if( (ref>0 and act>0) or (ref<=0 and act <=0)):
+            err_theta = ref-act
+        elif( ref<=0 and act > 0):
+            if(abs(ref-act)<abs(2*np.pi+ref-act)):
+                err_theta = -abs(act-ref)
+            else:
+                err_theta = abs(2*np.pi + ref- act)
+        else:
+            if(abs(ref-act)<abs(2*np.pi-ref+act)):
+                err_theta = abs(act-ref)
+            else: 
+                err_theta = -abs(2*np.pi-ref+act)
+
+
+        RotM = np.array([ 
+            [np.cos(-theta_current), -np.sin(-theta_current)],
+            [np.sin(-theta_current), np.cos(-theta_current)]
+        ])
+
+        errM = np.array([[ref_state_current[0]-x_current],[ref_state_current[1]-y_current]])
+
+        errRM = RotM@errM
+
+
+        error_state = [errRM[0][0],errRM[1][0],err_theta, ref_state_current[3]-v_current]
+
+        return error_state, ref_state_current[3]
+
+
     #callback to run a loop and publish data this class generates
     def pub_callback(self):
         if(not self.go):
             return
-        msg = Path()
+        if(self.planning_mode == "waypoints"):
+            msg = VehicleState()
+            error_state,ref_vel = self.wpts_path_plan()
+
+            msg.pose.position.x = error_state[0]
+            msg.pose.position.y = error_state[1]
+            msg.pose.orientation.z = error_state[2]
+            msg.twist.linear.x = error_state[3]
+            msg.twist.linear.y = ref_vel
+            self.pub_err_state.publish(msg)  
+
+        else:
+            msg = Path()
             
-        target_pt = self.plan_path()
-
-        #calculate path from current cone locations
-        if(self.vis):
-            plt.draw()
-            plt.pause(0.0001)
-
-        pt = PoseStamped()
-        pt.pose.position.x = target_pt[0]
-        pt.pose.position.y = target_pt[1]
-        msg.header.stamp = self.get_clock().now().to_msg()
-        msg.poses.append(pt)
-        self.pub_path.publish(msg)
+            target_pt = self.plan_path()
+
+            #calculate path from current cone locations
+            if(self.vis):
```

Comment:
as a note: () isn't needed in python if statements. might want to look into automatic python formatters like black for vscode. i'll also be adding an automatic formatter in the future when you do a commit, but worthwhile to always format your code.